### PR TITLE
Adds documentation for Access behaviour Callbacks

### PR DESCRIPTION
**Pull request by Qqwy**
This Pull Request adds documentation to the callbacks that need to be defined when implementing the Access behaviour. See #4937 .

I fully expect that some further changes are necessary before this can be merged.

Mayor points of concern:

- Why _is_ there actually a separate `get` callback? I tried thinking hard about a case where it couldn't be implemented using `fetch` but I was unable to come up with something. It has to have a reason; I hope someone can show an example, which also probably is useful in the documentation.
- I refer to implementations of Map and Keyword, which are the two implementations of the Access protocol that ship with Elixir (that I know of). Is this a good idea, or did I overdo it by adding it at every callback documentation?
- Is the documentation of `get_and_update` clear enough like this? It is a very complex function, so maybe it can be clarified further.
- Is the behavior of `get_and_update` supposed to be the same to `Map.get_and_update` when being passed `:pop` (i.e. removing the stored value instead of updating it) ?
- Does it need examples? Can we provide examples for callbacks?

### CODE REVIEW

**Comment by josevalim** [comment]
> Why is there actually a separate get callback? I tried thinking hard about a case where it couldn't be implemented using fetch but I was unable to come up with something.

Two reasons: 1. we expect the modules to define those so *others* can call it. 2. it keeps the API consistent (get and get_and_update)

> Is this a good idea, or did I overdo it by adding it at every callback documentation?

It is fine like this.

> Is the behavior of get_and_update supposed to be the same to Map.get_and_update when being passed :pop (i.e. removing the stored value instead of updating it) ?

Yes.

> Does it need examples? Can we provide examples for callbacks?

We already have examples in the `Access` functions, so I wouldn't add them.

Other than that, can you please make sure you wrap your docs around 80-100 columns.

File: lib/elixir/lib/access.ex [hunk 1]
```diff
@@ -113,15 +113,70 @@ defmodule Access do
...
+
+  Many of the functions defined in the `Access` module internally call this function.
+
+  See `Map.fetch/2` and `Keyword.fetch/2` for example implementations.
+
+  """
   @callback fetch(t, key) :: {:ok, value} | :error
+
+  @doc """
+  Invoked when someone wants to access the value stored under `key` in your structure. if `key` is not found, the second argument is returned.
```
Comment (eksperimental) [inline lib/elixir/lib/access.ex]: "if `key`" should read "If `key`"

File: lib/elixir/lib/access.ex [hunk 2]
```diff
@@ -113,15 +113,70 @@ defmodule Access do
...
+
+  This is the method that is used when someone calls `your_structure[some_key]`.
+
+  For some data structures, it might be possible to implement it by internally calling `fetch/2`, but for others this is not possible.
```
Comment (eksperimental) [inline lib/elixir/lib/access.ex]: :scissors: trailing space

File: lib/elixir/lib/access.ex [hunk 3]
```diff
@@ -113,15 +113,70 @@ defmodule Access do
...
+
+  For some data structures, it might be possible to implement it by internally calling `fetch/2`, but for others this is not possible. 
+  This is why both `fetch/2` and `get/3` exist.
+
+  See `Map.get/3` and `Keyword.get/3` for example implementations.
+  """
   @callback get(t, key, value) :: value
+
+  @doc """
+  Invoked when someone calls `Access.get_and_update/3`
```
Comment (eksperimental) [inline lib/elixir/lib/access.ex]: we need trailing period here.

File: lib/elixir/lib/access.ex [hunk 4]
```diff
@@ -113,15 +113,70 @@ defmodule Access do
...
+  
+  The implementation should invoke the passed function on the value inside key `key` in the passed structure.
+  This function will return a tuple in the form of `{received_value, new_value}`.
+  
+  The returned `new_value` should then be used to create an updated structure, where `new_value` is stored under `key` instead of its original value.
+  Finally, the implementation should return the tuple `{received_value, updated_structure}`
```
Comment (eksperimental) [inline lib/elixir/lib/access.ex]: missing trailing period

File: lib/elixir/lib/access.ex [hunk 5]
```diff
@@ -113,15 +113,70 @@ defmodule Access do
...
+
+  See `Map.get_and_update/3` or `Keyword.get_and_update/3` for example implementations.
+  """
   @callback get_and_update(t, key, (value -> {value, value} | :pop)) :: {value, t}
+  
+  @doc """
+  Invoked when someone calls `Access.pop/2`, and also used internally by some functions.
+
+  The implementation should remove `key` and whatever is inside it from the passed structure.
+  The result value should be the tuple `{removed_value, updated_structure}`
```
Comment (eksperimental) [inline lib/elixir/lib/access.ex]: missing trailing period.

**Comment by Qqwy** [comment]
Thank you, @josevalim and @eksperimental. I've added the suggested changes in the documentation.

File: lib/elixir/lib/access.ex [hunk 6]
```diff
@@ -113,15 +113,80 @@ defmodule Access do
...
   See the functions `key/1`, `key!/1`, `elem/1` and `all/0` for the current
   accessors.
+
+  ## Adding the Access behaviour to your own data structures
+  
+  If you want to be able to use the Access protocol 
+  with your own data structures, you can do so by adding
+
+  ```
+  @behaviour Access
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: Use four spaces to indent this insted.

File: lib/elixir/lib/access.ex [hunk 7]
```diff
@@ -113,15 +113,80 @@ defmodule Access do
 
   See the functions `key/1`, `key!/1`, `elem/1` and `all/0` for the current
   accessors.
+
+  ## Adding the Access behaviour to your own data structures
+  
+  If you want to be able to use the Access protocol 
+  with your own data structures, you can do so by adding
```
Comment (josevalim) [inline lib/elixir/lib/access.ex]: Please add `:` after "adding". :)

**Comment by josevalim** [comment]
@Qqwy I have added two final comments and we are good to go1

**Comment by josevalim** [comment]
:heart: :green_heart: :blue_heart: :yellow_heart: :purple_heart:

**Comment by josevalim** [comment]
@whatyouhide sorry, just now I saw you self-assigned yourself. :( If there are any issues left, please push them directly to master. :)

**Comment by Qqwy** [comment]
:heart_eyes: Thank you!

**Comment by whatyouhide** [comment]
@josevalim no worries! I just assigned myself so that you could choose not to review, but it's totally fine if you reviewed it :) 👍

**Pull request merged**